src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignedCmsTests.netcoreapp.cs

```diff
@@ -0,0 +1,498 @@
+
+            Assert.Equal(1, cms.SignerInfos.Count);
+
+            SignerInfo signer = cms.SignerInfos[0];
+            Assert.Equal(cert, signer.Certificate);
+
+            if (counterSignerCert != null)
+            {
+                Assert.Equal(1, signer.CounterSignerInfos.Count);
+                Assert.Equal(counterSignerCert, signer.CounterSignerInfos[0].Certificate);
+            }
+            else
+            {
+                Assert.Equal(0, signer.CounterSignerInfos.Count);
+            }
+        }
+
+        private static (byte[] signedData, X509Certificate2 pubCert) GetSignedData(byte[] content, Func<(X509Certificate2, AsymmetricAlgorithm)> getCertAndKey, Func<AsymmetricAlgorithm, bool> checkIfSignHashCalled)
+        {
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            (X509Certificate2 cert, AsymmetricAlgorithm key) = getCertAndKey();
+            using (key)
+            {
+                CmsSigner signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, cert, key);
+                if (key is DSA)
+                {
+                    signer.IncludeOption = X509IncludeOption.EndCertOnly;
+                    signer.DigestAlgorithm = new Oid(Oids.Sha1, Oids.Sha1);
+                }
+
+                cms.ComputeSignature(signer);
+
+                byte[] ret = cms.Encode();
+                Assert.True(checkIfSignHashCalled(key));
+
+                Assert.True(ByteArrayContains(ret, content));
+
+                return (ret, cert);
+            }
+        }
+
+        private static (byte[] counterSignedData, X509Certificate2 firstSignerCert, X509Certificate2 counterSignerCert) GetCounterSignedData(byte[] content, Func<(X509Certificate2, AsymmetricAlgorithm)> getFirstSignerCertAndKey, Func<(X509Certificate2, AsymmetricAlgorithm)> getCounterSignerCertAndKey, Func<AsymmetricAlgorithm, bool> signedHashCalled, Func<AsymmetricAlgorithm, bool> signedHashCalledForCounterSignerKey)
+        {
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            (X509Certificate2 firstSignerCert, AsymmetricAlgorithm firstSignerKey) = getFirstSignerCertAndKey();
+            using (firstSignerKey)
+            {
+                CmsSigner firstSigner = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, firstSignerCert, firstSignerKey);
+                if (firstSignerKey is DSA)
+                {
+                    firstSigner.IncludeOption = X509IncludeOption.EndCertOnly;
+                    firstSigner.DigestAlgorithm = new Oid(Oids.Sha1, Oids.Sha1);
+                }
+
+                cms.ComputeSignature(firstSigner);
+                Assert.True(signedHashCalled(firstSignerKey));
+            }
+
+            byte[] signed = cms.Encode();
+            Assert.True(ByteArrayContains(signed, content));
```

Comment:
This seems like over-testing.  But if you're concerned, I'd recommend just using Encode then calling Decode on a new instance.  If the content got lost the signature will fail.